File: ios/MendixNative/DevAppMenu/DevAppMenu.m
#import "DevAppMenu.h"
#import <UIKit/UIKit.h>
#import <React/RCTBridge.h>
#import <React/RCTUtils.h>
#import "React/RCTRedBox.h"
#import "ReactNative.h"
#import "AppMenuProtocol.h"
#import "AppPreferences.h"
#import "UIAlertControllerExt.h"
#import "UIAlertActionExt.h"
#import "DevAppMenuUIAlertController.h"

@implementation DevAppMenu

typedef void (^ ShowAlertHandler)(UIAlertAction*);

ShowAlertHandler showAlertHandler;
ShowAlertHandler showAdvancedAlertHandler;

- (void) show:(BOOL)devMode {
    RCTBridge *bridge = [[ReactNative instance] getBridge];
    UIWindow *window = UIApplication.sharedApplication.keyWindow;

    UIAlertControllerStyle style = UIDevice.currentDevice.userInterfaceIdiom == UIUserInterfaceIdiomPhone ? UIAlertControllerStyleActionSheet : UIAlertControllerStyleAlert;
    UIAlertController *alert = [DevAppMenuUIAlertController alertControllerWithTitle:@"App menu" message:nil preferredStyle:style];
    UIAlertController *advanceAlert = [DevAppMenuUIAlertController alertControllerWithTitle:@"Advance settings" message:nil preferredStyle:style];
    
    showAlertHandler = [self createShowAlert:alert completion:nil];
    showAdvancedAlertHandler = [self createShowAlert:advanceAlert completion: ^{
        [advanceAlert applyAccessibilityIdentifiers];
    }];
    
    if (devMode) {
        [self addDevModeAction:alert advancedAlert: advanceAlert];
    }
    
    UIAlertAction *reloadAction = [UIAlertAction actionWithTitle:@"Refresh" style:UIAlertActionStyleDefault handler:^void(UIAlertAction * _Nonnull action) {
        [[ReactNative instance] reload];
    }];
    [reloadAction setAccessibilityIdentifier: @"reload_button"];
    [alert addAction:reloadAction];

    UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Return To Homescreen" style:UIAlertActionStyleDefault handler:^void(UIAlertAction * _Nonnull action) {
        if (bridge.redBox) {
            [bridge.redBox dismiss];
        }
        [[ReactNative instance] stop];
    }];
    [closeAction setAccessibilityIdentifier: @"close_button"];
    [alert addAction:closeAction];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDefault handler:nil];
    [cancelAction setAccessibilityIdentifier: @"cancel_button"];
    [alert addAction:cancelAction];
    
    if (RCTPresentedViewController() == nil) {
        if (window != nil) {
            window.rootViewController = [[UIViewController alloc] initWithNibName:nil bundle:nil];
        }
    }
    
    if ([RCTPresentedViewController() class] != [UIAlertController class]) {
        [RCTPresentedViewController() presentViewController:alert animated:YES completion:^{
            [alert applyAccessibilityIdentifiers];
        }];
    }
}

- (void) addDevModeAction:(UIAlertController *)alert advancedAlert:(UIAlertController *)advancedAlert {
    BOOL isDebuggingRemotely = [[ReactNative instance] isDebuggingRemotely];
    
    [self addAdvanceSettingsAction:alert advancedAlert:advancedAlert];
    
    UIAlertAction *advancedSettingsAction = [UIAlertAction actionWithTitle:@"Advanced settings" style:UIAlertActionStyleDefault handler:showAdvancedAlertHandler];
    [advancedSettingsAction setAccessibilityIdentifier: @"advanced_settings_button"];
    [alert addAction:advancedSettingsAction];
    
    NSString *remoteDebuggingTitle = [(!isDebuggingRemotely ? @"Enable" : @"Disable") stringByAppendingString:@" remote JS debugging"];
    UIAlertAction *remoteDebuggingAction = [UIAlertAction actionWithTitle:remoteDebuggingTitle style:UIAlertActionStyleDefault handler:^void(UIAlertAction * _Nonnull action) {
        [[ReactNative instance] remoteDebugging:!isDebuggingRemotely];
    }];
    [remoteDebuggingAction setAccessibilityIdentifier: @"remote_debugging_button"];
    [alert addAction:remoteDebuggingAction];
    
    UIAlertAction *toggleElementInspectorAction = [UIAlertAction actionWithTitle:@"Toggle Element Inspector" style:UIAlertActionStyleDefault handler:^void(UIAlertAction * _Nonnull action) {
        [AppPreferences setElementInspector:![AppPreferences isElementInspectorEnabled]];
        [[ReactNative instance] toggleElementInspector];
    }];
    [toggleElementInspectorAction setAccessibilityIdentifier: @"toggle_inspector_button"];
    [alert addAction:toggleElementInspectorAction];
}

- (void) addAdvanceSettingsAction:(UIAlertController *)alert advancedAlert:(UIAlertController *)advancedAlert {
    UIAlertAction *clearDataButtonAction = [UIAlertAction actionWithTitle:@"Clear Data" style:UIAlertActionStyleDestructive handler:^void(UIAlertAction * _Nonnull action) {
        [[ReactNative instance] clearData];
        [[ReactNative instance] reload];
    }];
    [clearDataButtonAction setAccessibilityIdentifier: @"clear_data_button"];
    [clearDataButtonAction setEnabled:[[ReactNative instance] getBridge] != nil];
    [advancedAlert addAction:clearDataButtonAction];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Back" style:UIAlertActionStyleCancel handler:showAlertHandler];
    [cancelAction setAccessibilityIdentifier: @"cancel_button"];
    [advancedAlert addAction:cancelAction];
}

- (ShowAlertHandler) createShowAlert:(UIAlertController *)alert completion:(void (^)(void)) completion {
    return ^(UIAlertAction *action) {
        [RCTPresentedViewController() presentViewController:alert animated:YES completion:completion];
    };
}

@end

File: ios/Modules/DevAppMenu/DevAppMenu.swift
//import UIKit
//import React
//import React_RCTAppDelegate
//
//@objc class DevAppMenu: NSObject, AppMenuProtocol {
//    
//    // MARK: - Type Aliases
//    typealias ShowAlertHandler = (UIAlertAction) -> Void
//    
//    // MARK: - Private Properties
//    private var showAlertHandler: ShowAlertHandler?
//    private var showAdvancedAlertHandler: ShowAlertHandler?
//    
//    // MARK: - AppMenuProtocol Implementation
//    @objc func show(_ devMode: Bool) {
////        let redbox = RedBoxHelper.shared.redBox
//        
//        let window = UIApplication.shared.connectedScenes
//            .first {$0.activationState == .foregroundActive}
//            .flatMap{ $0 as? UIWindowScene }?
//            .windows
//            .first { $0.isKeyWindow }
//        
//        let style: UIAlertController.Style = UIDevice.current.userInterfaceIdiom == .phone ? .actionSheet : .alert
//        let alert = DevAppMenuUIAlertController(title: "App menu", message: nil, preferredStyle: style)
//        let advanceAlert = DevAppMenuUIAlertController(title: "Advance settings", message: nil, preferredStyle: style)
//        
//        showAlertHandler = createShowAlert(alert, completion: nil)
//        showAdvancedAlertHandler = createShowAlert(advanceAlert) {
//            advanceAlert.applyAccessibilityIdentifiers()
//        }
//        
//        if devMode {
//            addDevModeAction(alert, advancedAlert: advanceAlert)
//        }
//        
//        // Refresh Action
//        let reloadAction = UIAlertAction(title: "Refresh", style: .default) { _ in
//            ReactNative.instance.reload()
//        }
//        reloadAction.setAccessibilityIdentifier("reload_button")
//        alert.addAction(reloadAction)
//        
//        // Return To Homescreen Action
////        let closeAction = UIAlertAction(title: "Return To Homescreen", style: .default) { _ in
////            redbox.dismiss()
////            ReactNative.instance.stop()
////        }
////        closeAction.setAccessibilityIdentifier("close_button")
////        alert.addAction(closeAction)
////        
//        // Cancel Action
//        let cancelAction = UIAlertAction(title: "Cancel", style: .default, handler: nil)
//        cancelAction.setAccessibilityIdentifier("cancel_button")
//        alert.addAction(cancelAction)
//        
//        // Present the alert
//        
//        let presentedVC = RCTPresentedViewController()
//        
//        if presentedVC == nil {
//            if let window {
//                window.rootViewController = UIViewController(nibName: nil, bundle: nil)
//            }
//        }
//        
//        if type(of: presentedVC) != UIAlertController.self {
//            presentedVC?.present(alert, animated: true) {
//                alert.applyAccessibilityIdentifiers()
//            }
//        }
//    }
//    
//    // MARK: - Private Methods
//    private func addDevModeAction(_ alert: UIAlertController, advancedAlert: UIAlertController) {
//        let isDebuggingRemotely = ReactNative.instance.isDebuggingRemotely()
//        
//        addAdvanceSettingsAction(alert, advancedAlert: advancedAlert)
//        
//        // Advanced Settings Action
//        let advancedSettingsAction = UIAlertAction(title: "Advanced settings", style: .default, handler: showAdvancedAlertHandler)
//        advancedSettingsAction.setAccessibilityIdentifier("advanced_settings_button")
//        alert.addAction(advancedSettingsAction)
//        
//        // Remote Debugging Action
//        let remoteDebuggingTitle = "\(isDebuggingRemotely ? "Disable" : "Enable") remote JS debugging"
//        let remoteDebuggingAction = UIAlertAction(title: remoteDebuggingTitle, style: .default) { _ in
//            ReactNative.instance.remoteDebugging(!isDebuggingRemotely)
//        }
//        remoteDebuggingAction.setAccessibilityIdentifier("remote_debugging_button")
//        alert.addAction(remoteDebuggingAction)
//        
//        // Toggle Element Inspector Action
//        let toggleElementInspectorAction = UIAlertAction(title: "Toggle Element Inspector", style: .default) { _ in
//            AppPreferences.elementInspectorEnabled.toggle()
//            ReactNative.instance.toggleElementInspector()
//        }
//        toggleElementInspectorAction.setAccessibilityIdentifier("toggle_inspector_button")
//        alert.addAction(toggleElementInspectorAction)
//    }
//    
//    private func addAdvanceSettingsAction(_ alert: UIAlertController, advancedAlert: UIAlertController) {
//        // Clear Data Action
//        let clearDataButtonAction = UIAlertAction(title: "Clear Data", style: .destructive) { _ in
//            ReactNative.instance.clearData()
//            ReactNative.instance.reload()
//        }
//        clearDataButtonAction.setAccessibilityIdentifier("clear_data_button")
//        clearDataButtonAction.isEnabled = RCTBridge.current() != nil
//        advancedAlert.addAction(clearDataButtonAction)
//        
//        // Back Action
//        let cancelAction = UIAlertAction(title: "Back", style: .cancel, handler: showAlertHandler)
//        cancelAction.setAccessibilityIdentifier("cancel_button")
//        advancedAlert.addAction(cancelAction)
//    }
//    
//    private func createShowAlert(_ alert: UIAlertController, completion: (() -> Void)?) -> ShowAlertHandler {
//        return { _ in
//            RCTPresentedViewController()?.present(alert, animated: true, completion: completion)
//        }
//    }
//}
//
////Checked
